src/common/walFilter/versions/recordProcessGPDB7.c

```diff
@@ -0,0 +1,436 @@
+            COPY_HEADER_FIELD(&size, sizeof(size));
+            mainDataSize = size;
+            break;              /* by convention, the main data fragment is
+                                 * always last */
+        }
+        else if (block_id == XLR_BLOCK_ID_ORIGIN)
+        {
+            ptr += sizeof(RepOriginId);
+            remaining -= (uint32_t) sizeof(RepOriginId);
+        }
+        else if (block_id <= XLR_MAX_BLOCK_ID)
+        {
+            /* XLogRecordBlockHeader */
+            uint8 fork_flags;
+
+            if (block_id <= maxBlockId)
+            {
+                THROW_FMT(FormatError, "out-of-order block_id %u", block_id);
+            }
+            maxBlockId = block_id;
+
+            COPY_HEADER_FIELD(&fork_flags, sizeof(uint8));
+            bool has_image = ((fork_flags & BKPBLOCK_HAS_IMAGE) != 0);
+            bool has_data = ((fork_flags & BKPBLOCK_HAS_DATA) != 0);
+
+            uint16_t data_len;
+
+            COPY_HEADER_FIELD(&data_len, sizeof(uint16));
+            /* cross-check that the HAS_DATA flag is set iff data_length > 0 */
+            if (has_data && data_len == 0)
+            {
+                THROW_FMT(FormatError, "BKPBLOCK_HAS_DATA set, but no data included");
+            }
+            if (!has_data && data_len != 0)
+            {
+                THROW_FMT(FormatError, "BKPBLOCK_HAS_DATA not set, but data length is %u", (unsigned int) data_len);
+            }
+            datatotal += data_len;
+
+            if (has_image)
+            {
+                uint16_t bimg_len;
+                uint16_t hole_offset;
+                uint8 bimg_info;
+
+                COPY_HEADER_FIELD(&bimg_len, sizeof(uint16));
+                COPY_HEADER_FIELD(&hole_offset, sizeof(uint16));
+                COPY_HEADER_FIELD(&bimg_info, sizeof(uint8));
+
+                uint16_t hole_length;
+                if (bimg_info & BKPIMAGE_IS_COMPRESSED)
+                {
+                    if (bimg_info & BKPIMAGE_HAS_HOLE)
+                        COPY_HEADER_FIELD(&hole_length, sizeof(uint16));
+                    else
+                        hole_length = 0;
+                }
+                else
+                    hole_length = (uint16_t) (pageSize - bimg_len);
+                datatotal += bimg_len;
+
+                /*
+                 * cross-check that hole_offset > 0, hole_length > 0 and
+                 * bimg_len < BLCKSZ if the HAS_HOLE flag is set.
+                 */
+                if ((bimg_info & BKPIMAGE_HAS_HOLE) &&
+                    (hole_offset == 0 ||
+                     hole_length == 0 ||
+                     bimg_len == pageSize))
+                {
+                    THROW_FMT(FormatError, "BKPIMAGE_HAS_HOLE set, but hole offset %u length %u block image length %u",
+                              (unsigned int) hole_offset,
+                              (unsigned int) hole_length,
+                              (unsigned int) bimg_len);
```

Comment:
Let's avoid casting here and below
```suggestion
                    THROW_FMT(FormatError, "BKPIMAGE_HAS_HOLE set, but hole offset %" PRIu16 " length %" PRIu16 " block image length %" PRIu8,
                              hole_offset,
                              hole_length,
                              bimg_len);
```